[클린코드 4기 오윤구] 계산기 구현

oyg0420 opened the pull request:
## What
- 계산기 기능 구현 및 Cypress 적용한 테스트 코드 작성

## How
- Cypress
   - Cypress Custom Command 작성
   - 기능 구현 사항 테스트 코드 작성
- Calculator
  - 숫자, 연산자, 연산 결과값 및 관리 함수 작성
  - 연산 함수 작성: 덧셈, 뺄샘, 나눗셈, 곱셈, 할당
  - 검증 함수 작성: 최대 3자리 수 
- App
  - 이벤트 핸들러 등록: 숫자, 연산자, AC 버튼
  - html update 함수 구현

## 느낀점
React로만 작업하다 라이브러리 의존없이 작업하려니 시작부터 고민했습니다  ㅎㅎ 
고민한 부분은 역할 분리를 하기 위한 고민이 가장 컸습니다. Calculator 클래스는 계산기 관련 집합체 역할, App 클래스는 이벤트 등록과 화면을 업데이트하는 역할로 구분했습니다. 더 구분할 수 있는 부분과 방법에 있다면 리뷰어님이 확인해주시고 피드백 부탁드립니다. 
Cypress도 적용해보니 테스트 통과하는 쾌감이 매우 좋더군요! 앞으로 Cypress 관련 공부도 많이 필요할것 같아요.
주말 동안 재밌게 작업했습니다. 감사합니다

oyg0420 committed:
style: lint 적용

1. html, css lint 적용

oyg0420 committed:
chore: cypress intall

1. cypress 기본 설정: video disabled
2. calculator test code todo 작성

oyg0420 committed:
chore: live-server install

1. live-server install 및 script 작성

oyg0420 committed:
feat: calculator class 생성

1. 각 연산자 로직 정의
2. const: 기본값, 연산자 관리

oyg0420 committed:
feat: calculator 로직 정리

1. private field 생성
2. getDisplay: 계산기 노출 값
3. validation 함수 생성
4. const 추가: '='

oyg0420 committed:
feat: calculator operation test code

1. 사칙연산 관련 테스트 코드 작성

oyg0420 committed:
feat: utils 함수 추가

1. isNumber: 숫자형 검증 함수 추가
2. getAppendedNumber: calculator 메서드 이름 변경

oyg0420 committed:
feat: calculator 테스트 코드 추가 작성

1. cypress custom command 추가
2. custom command 적용

oyg0420 committed:
feat: app class 생성

1. initEventHandler: 계산기 조작 eventHandler 적용
2. updateDisplay: 계산기 조작 시 html display 되도록 적용

oyg0420 committed:
refactor: calculator instance 구조분해할당

1. calculator class 내부 method arrow function으로 변경: this -> Calculator
2. setOperator 메서드: #operator 존재한 경우에는 할당하지 않도록 변경

oyg0420 committed:
chore: file name 수정

oyg0420 committed:
feat: display 로직 변경

1. 숫자만 노출되도록 변경
2. #num* 기본값 null 변경: 입력 상태 구분

oyg0420 committed:
refactor: isOverMaxLength 이관

1. isOverMaxLength -> isOver 함수로 변경

EungyuCho commented on cypress/e2e/calculator.cy.js:
```diff
@@ -0,0 +1,72 @@
+    cy.clickOperator(OPERATOR.SUM);
+    cy.clickDigit(3);
+    cy.clickOperator(OPERATOR.ASSIGNMENT);
+    cy.getTotal('8');
```
(마이너) digit는 number타입으로 받는데 getTotal함수는 문자열로 받을 이유가 있을까요? 👀

oyg0420 replied:
문자열로 받을 특별한 이유는 없습니다. 😅 🙏

EungyuCho commented on src/js/calculator.js:
```diff
@@ -0,0 +1,74 @@
+  };
+
+  compute = () => {
+    this.#output = OPERRATIONS[this.#operator](this.#num1, this.#num2);
```
```suggestion
    const operation = OPERRATIONS[this.#operator];
    this.#output = operation(this.#num1, this.#num2);
```
개인적으로는 이렇게 함수에 네이밍을 붙혀서 재선언하고 사용하는게 더 이해하기 쉽더라구요 🙂

oyg0420 replied:
짧게 작성한다고 좋은 코드가 아닌거 같네요 ㅎㅎ 제안해주신것처럼 재선언하는 방식이 더욱 읽기 편해진것 같아요 ! !

EungyuCho commented on src/js/calculator.js:
```diff
@@ -0,0 +1,74 @@
+      this.#num2 = nextNumber;
+    } else {
+      this.#num1 = nextNumber;
+    }
```
조건문에는 boolean 타입을 넣는것을 권장드려요! 
더 명확하게 조건을 표기하는게 좋다고 생각합니다~ [관련 eslint rule](https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/docs/rules/strict-boolean-expressions.md)

EungyuCho commented on src/js/calculator.js:
```diff
@@ -0,0 +1,74 @@
+  #operator;
+
+  constructor() {
+    this.#num1 = null;
+    this.#num2 = null;
+    this.#output = null;
+    this.#operator = null;
```
생성자에서 null로 초기화 한 이유가 있을까요??
초기화를 안하는 것과 어떤 차이가 있을까요? 👀

oyg0420 replied:
1. 생성자에서 null로 초기화 한 이유가 있을까요?
각 멤버들이 조건문에서 사용될때 값의 존재 유무를 null를 기준만으로 파악할 수 있어 사용했습니다.
숫자형으로 초기화를 생각했지만 값으로 표현해야할 0도 falsy 값이기 때문에 조건문으로 적당하다고 생각하지 않았습니다.

2. 초기화를 안하는 것과 어떤 차이가 있을까요?
초기화 하지 않으면 멤버 변수는 undefined을 갖게 됩니다. 추가로 자바스크립트엔진은 아래 같은 경우에 undefined을 반환합니다.
- 변수에 값을 대입하지 않은 경우
- 함수에서 return 문이 없는 경우
- 객체에서 존재하지 않은 property 접근하는 경우

물론 위와 같은 경우 외에도 사용자가 직접 undefiend를 변수에 할당할 수 있겠지만, 동일하게 **비어있음**을 나타내는 null를 할당한다면, **undefined는 자바크립트엔진이 반환해주는 비어있는 값, null은 사용자가 비어있음을 나타내는 값으로 구분**하여 인지하는데 도움이 될 것 같습니다.

EungyuCho replied:
확실히 숫자로 초기화해줬다면 계산기의 Display를 출력하는데 더 까다로웠을 것 같아요. 
null에 대한 근거가 명확해서 좋네요! undefined에 대한 설명도 완벽했습니다~ 💯

EungyuCho commented on src/js/calculator.js:
```diff
@@ -0,0 +1,74 @@
+
+    if (this.#operator) {
+      currentNumber = this.#num2 ?? DEFAULT_NUMBER;
+    }
```
```suggestion
    const currentNumber = (this.#operator === null ?  this.#num1 : this.#num2) ?? DEFAULT_NUMBER;
```

EungyuCho commented on src/js/utils.js:
```diff
@@ -0,0 +1,14 @@
+
+export function isNodeContains(parentNode, targetNode) {
+  return parentNode.contains(targetNode);
+}
```
이 함수는 어떤 이유로 분리한지 알 수 있을까요? 🤔

oyg0420 replied:
queryselector를 통해 얻은 노드에 타켓이 포함이 되었는지 확인하기 위해서 분리했지만..굳이 분리할 필요성은 못느끼겠네요.

EungyuCho commented on src/js/utils.js:
```diff
@@ -0,0 +1,14 @@
+export function isNodeContains(parentNode, targetNode) {
+  return parentNode.contains(targetNode);
+}
+export function isOver(max, target) {
+  if (target.length > max) {
+    return true;
+  }
+
+  return false;
+}
```
메서드명은 구체적일수록 좋다고 생각해요. isOver이 length를 체크한다고 인지하기 어려울 수도 있겠다고 생각했어요.

그리고 target.length > max는 boolean을 리턴하니 한줄로 리턴해버리면 더 좋을 것 같습니다 :)

oyg0420 replied:
와우. 간단하게 한줄로 반환할 수 있었네요. 😅

EungyuCho commented on src/js/calculator.js:
```diff
@@ -0,0 +1,74 @@
+import { isNumber } from './utils.js';
+import { DEFAULT_NUMBER, OPERATOR } from './const.js';
+
+const OPERRATIONS = {
```
저도 변수명을 지을 때 오타를 많이 치곤해서 이러한 [extention](https://marketplace.visualstudio.com/items?itemName=streetsidesoftware.code-spell-checker)을 활용하고있어요

oyg0420 replied:
오... 감사합니다. 바로 사용했습니다. 🙏

EungyuCho commented on src/js/app.js:
```diff
@@ -0,0 +1,71 @@
+  }
+
+  handleClickTarget = (e) => {
+    if (isNodeContains(this.querySelector('.digits'), e.target)) {
+      this.handleClickDigit(e.target.textContent);
+      this.updateDisplay();
+    }
+
+    if (isNodeContains(this.querySelector('.modifiers'), e.target)) {
+      this.handleClickAllClear();
+      this.updateDisplay();
+    }
+
+    if (isNodeContains(this.querySelector('.operations'), e.target)) {
+      this.handleClickOperator(e.target.textContent);
+      this.updateDisplay();
+    }
```
이와 같은 문제점들이 보여요.
1. 클릭을 할때마다 모든 조건문을 검사해요. (숫자를 눌렀으면 이후 조건문을 체크할 필요 X)
2. 1번 문제에 의해서 클릭을 할 때마다 query selector를 3번씩 진행해요

1번 문제는 2번 문제를 해결하면 딱히 수정할 필요가 없을 것 같아요.
queryselector를 매번 실행하는게 아니라 미리 어떤 노드들이 있는지 알고있으면 해결할 수 있는 문제로 보여요.
이번 미션에서는 UI가 바뀔 여지가 없을 것 같은데 그렇다면 생성자에서 미리 선택할 노드들을 변수에 할당해보면 어떨까요?

oyg0420 replied:
말씀해주신 부분에 동의합니다.
1. 클릭 이벤트 모든 조건문 실행
2. 매번 queryselector를 실행
ㅜ . ㅜ  감사합니다.

EungyuCho commented on src/js/app.js:
```diff
@@ -0,0 +1,71 @@
+  updateDisplay() {
+    const { getDisplay } = this.calculator;
+
+    this.$target.querySelector('#total').innerHTML = getDisplay();
```
내부 텍스트만 변경하고 있어서 innerText로 해도 될거같아요

EungyuCho reviewed (changes requested):
윤구님 고생하셨습니다~

모듈의 분리에 대해서는 피드백을 추가로 안드려도 될 것 같아요! 깔끔하게 분리 된 것 같습니다 :)
클린코드에 대한 부분에서 조금 더 개선해볼만한 여지가 있는 부분들이 있어 코멘트를 남겨드렸어요

나중에 확인해보시고 변경할만한 부분이 있으시면 변경부탁드릴게요!

oyg0420 committed:
refactor: 코드리뷰 반영

1. 오타수정
2. 불필요 함수 제거: isNodeContains
3. 함수 naming 수정: isOver
4. 초기화시 queryselector 사용하여 변수 할당($digits, $modifiers, $operations, $total)
5. 이벤트 핸들러 if block 문 return 추가

EungyuCho commented on src/js/app.js:
```diff
@@ -0,0 +1,74 @@
+
+    const nextNumber = appendNumber(Number(digit));
+    if (isOverMaxLength(MAX_LENGTH, nextNumber.toString())) {
+      alert('숫자는 한번에 최대 3자리 수까지 입력 가능합니다.');
```
(매우 사소) 3자리 => ${MAX_LENGTH}자리로 해두면 좋을 것 같아요

EungyuCho commented on src/js/app.js:
```diff
@@ -0,0 +1,74 @@
+    }
+  }
+
+  handleClickTarget = (e) => {
+    if (this.$digits.contains(e.target)) {
+      this.handleClickDigit(e.target.textContent);
+      this.updateDisplay();
+      return;
+    }
+
+    if (this.$modifiers.contains(e.target)) {
+      this.handleClickAllClear();
+      this.updateDisplay();
+      return;
+    }
+
+    if (this.$operations.contains(e.target)) {
+      this.handleClickOperator(e.target.textContent);
+      this.updateDisplay();
+      return;
+    }
+  };
+
+  initEventHandler() {
+    this.$target.addEventListener('click', this.handleClickTarget);
+  }
```
```suggestion
  handleClickCalculatorButton = (e) => {
    if (this.$digits.contains(e.target)) {
      this.handleClickDigit(e.target.textContent);
      return;
    }

    if (this.$modifiers.contains(e.target)) {
      this.handleClickAllClear();
      return;
    }

    if (this.$operations.contains(e.target)) {
      this.handleClickOperator(e.target.textContent);
    }
  };

  initEventHandler() {
    this.$target.addEventListener('click', (event) => {
      this.handleClickCalculatorButton(event);
      this.updateDisplay();
    }
  }
```
`updateDisplay`를 분리하는편이 더 좋을 것 같아요.

EungyuCho reviewed (approved):
안녕하세요 윤구님~
반영해주신 피드백 모두 확인했습니다 

미션을 진행하면서 commit과 메세지까지 상세하게 작업내용이 적혀있었고 주제인 TDD에 맞게 작업도 진행해주셔서 좋았어요. 👏

이미 코드도 충분히 깔끔해져서.. Approve 하겠습니다~ 
다음 미션도 화이팅입니다 :)
